packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -0,0 +1,314 @@
+
+  final _exceptionPauseMode = ValueNotifier<String>(null);
+  ValueListenable<String> get exceptionPauseMode => _exceptionPauseMode;
+
+  InstanceRef _reportedException;
+
+  void setVmService(VmService service) {
+    _service = service;
+    switchToIsolate(serviceManager.isolateManager.selectedIsolate);
+
+    _debugSubscription = _service.onDebugEvent.listen(_handleIsolateEvent);
+    serviceManager.isolateManager.onSelectedIsolateChanged
+        .listen(switchToIsolate);
+  }
+
+  void switchToIsolate(IsolateRef ref) async {
+    isolateRef = ref;
+
+    _isPaused.value = false;
+
+    _clearCaches();
+
+    if (ref == null) {
+      _breakpoints.value = [];
+      return;
+    }
+
+    final dynamic result = await _service.getIsolate(isolateRef.id);
+    if (result is Isolate) {
+      final Isolate isolate = result;
+
+      if (isolate.pauseEvent != null &&
+          isolate.pauseEvent.kind != EventKind.kResume) {
+        lastEvent = isolate.pauseEvent;
+        _reportedException = isolate.pauseEvent.exception;
+        _isPaused.value = true;
+      }
+
+      _breakpoints.value = isolate.breakpoints;
+
+      _exceptionPauseMode.value = isolate.exceptionPauseMode;
+    }
+  }
+
+  Future<Success> pause() => _service.pause(isolateRef.id);
+
+  Future<Success> resume() => _service.resume(isolateRef.id);
+
+  Future<Success> stepOver() {
+    // Handle async suspensions; issue StepOption.kOverAsyncSuspension.
+    final bool useAsyncStepping = lastEvent?.atAsyncSuspension == true;
+    return _service.resume(isolateRef.id,
+        step: useAsyncStepping
+            ? StepOption.kOverAsyncSuspension
+            : StepOption.kOver);
+  }
+
+  Future<Success> stepIn() {
+    return _service.resume(isolateRef.id, step: StepOption.kInto);
+  }
+
+  Future<Success> stepOut() {
+    return _service.resume(isolateRef.id, step: StepOption.kOut);
+  }
+
+  Future<void> clearBreakpoints() async {
+    final List<Breakpoint> breakpoints = _breakpoints.value.toList();
+    await Future.forEach(breakpoints, (Breakpoint breakpoint) {
+      return removeBreakpoint(breakpoint);
+    });
+  }
+
+  Future<void> addBreakpoint(String scriptId, int line) {
+    return _service.addBreakpoint(isolateRef.id, scriptId, line);
+  }
+
+  Future<void> addBreakpointByPathFragment(String path, int line) async {
+    final ScriptRef ref =
+        scripts.firstWhere((ref) => ref.uri.endsWith(path), orElse: () => null);
+    if (ref != null) {
+      return _service.addBreakpoint(isolateRef.id, ref.id, line);
+    }
+  }
+
+  Future<void> removeBreakpoint(Breakpoint breakpoint) {
+    return _service.removeBreakpoint(isolateRef.id, breakpoint.id);
+  }
+
+  Future<void> setExceptionPauseMode(String mode) {
+    return _service.setExceptionPauseMode(isolateRef.id, mode);
+  }
+
+  Future<Stack> getStack() async {
+    final stack = await _service.getStack(isolateRef.id);
+    if (stack is Sentinel) return null;
```

Comment:
I thought the VM Service apis changed to throw exceptions rather than return sentinels for these cases with the latest 4.0 version. Was this case missed?